### ISSUE TITLE
fs with kvision-electron

### ISSUE THREAD

**Issue by Lyeeedar** [rjaros/kvision#172]
I'm building off the kvision electron template, and trying to read a files contents. There seem to be interfaces for accessing all sorts of things (like dialogs, single instance applications, the shell, etc) but nothing for reading / writing to disk.

I tried using fs from the node-ket library, as this little code snippet shows:
```
import node.fs.fs

val remote: Remote = require("electron").remote
val file = remote.dialog.showOpenDialog(remote.getCurrentWindow(), object : OpenDialogOptions{})

val contents = fs.readFileStringSync(file, "utf8")
```

but that gives me the build error When accessing module declarations from UMD, they must be marked by both @JsModule and @JsNonModule

Entire ElectronApp.kt file with this code in (just drop into the kvision template-electron to try it):

```
package com.example

import node.fs.fs
import pl.treksoft.kvision.Application
import pl.treksoft.kvision.electron.BrowserWindow
import pl.treksoft.kvision.electron.OpenDialogOptions
import pl.treksoft.kvision.electron.Remote
import pl.treksoft.kvision.electron.dialog
import pl.treksoft.kvision.electron.nodejs.Process
import pl.treksoft.kvision.html.button
import pl.treksoft.kvision.html.div
import pl.treksoft.kvision.i18n.DefaultI18nManager
import pl.treksoft.kvision.i18n.I18n
import pl.treksoft.kvision.i18n.I18n.tr
import pl.treksoft.kvision.panel.FlexAlignItems
import pl.treksoft.kvision.panel.hPanel
import pl.treksoft.kvision.panel.root
import pl.treksoft.kvision.panel.vPanel
import pl.treksoft.kvision.require
import pl.treksoft.kvision.startApplication
import pl.treksoft.kvision.utils.createInstance
import pl.treksoft.kvision.utils.obj
import pl.treksoft.kvision.utils.px
import kotlin.browser.window

external val process: Process

class ElectronApp : Application() {

    val remote: Remote = require("electron").remote
    private var nativeWindow: BrowserWindow? = null

    override fun start() {

        I18n.manager =
            DefaultI18nManager(
                mapOf(
                    "en" to require("i18n/messages-en.json"),
                    "pl" to require("i18n/messages-pl.json")
                )
            )

        root("kvapp") {
            vPanel(alignItems = FlexAlignItems.CENTER, spacing = 30) {
                marginTop = 50.px
                fontSize = 30.px
                hPanel(spacing = 20) {
                    div(tr("Electron version"))
                    div("${process.versions?.electron}")
                    button("open file") {
                        onClick {
                            val remote: Remote = require("electron").remote
                            val file = remote.dialog.showOpenDialog(remote.getCurrentWindow(), object : OpenDialogOptions{})
                            val contents = fs.readFileStringSync(file, "utf8")
                            this.text = contents
                        }
                    }
                }
                hPanel(spacing = 20) {
                    div(tr("Chrome version"))
                    div("${process.versions?.chrome}")
                }
            }
        }

        nativeWindow = remote.BrowserWindow.createInstance(obj {
            title = "Native window"
            width = 700
            height = 500
        })
        nativeWindow?.on("closed")
        { _ ->
            nativeWindow = null
        }

        window.onunload = {
            nativeWindow?.destroy()
            nativeWindow = null
            Unit
        }
    }

    override fun dispose(): Map<String, Any> {
        super.dispose()
        nativeWindow?.destroy()
        nativeWindow = null
        return mapOf()
    }
}

fun main() {
    startApplication(::ElectronApp)
}
```

Any help on getting this to work would be fantastic.

**Comment by rjaros** [comment]
The easiest way to fix the modules problem is to change:
```
    moduleKind = "umd"
```
to
```
    moduleKind = "commonjs"
```
in your `build.gradle.kts` file.

**Comment by rjaros** [comment]
BTW I see that Electron API wrappers in kvision-electron module are somehow outdated. 
I had to use the following code to make open file dialog work:
```
                    button("open file") {
                        onClick {
                            val remote: Remote = require("electron").remote
                            val paths = remote.dialog.asDynamic().showOpenDialogSync(obj {})
                            if (paths != undefined) {
                                val contents = fs.readFileStringSync(paths[0], "utf8")
                                this.text = contents
                            }
                        }
                    }
```
Please leave this issue open. I'll try to regenerate wrappers from the current Electron version.

**Comment by Lyeeedar** [comment]
Thank you, that worked.

How easy would it be to add support for fs to the electron kvision? As one of the main reasons to use an electron app over a website is to interact with the disk, and the node.fs library only exposes a couple of the fs interfaces.

I would be happy to submit a pull request, if you can give some direction on what is required to add those wrappers?

**Comment by rjaros** [comment]
I've regenerated Electron bindings from the current Electron TypeScript definitions using Dukat. It's a lot (!) of code (20k lines of Kotlin code, 40 packages, hundreds of functions and classes). I have been able to test only some of them, but they seem to work fine. I'll release this with new KVision version soon. 

The `require` calls, `node` package or dynamic types will be no more needed for the same functionality. The application code will look like this:
```
package com.example

import pl.treksoft.kvision.Application
import pl.treksoft.kvision.electron.Electron.BrowserWindow
import pl.treksoft.kvision.electron.Electron.BrowserWindowConstructorOptions
import pl.treksoft.kvision.electron.Electron.OpenDialogSyncOptions
import pl.treksoft.kvision.electron.Electron.remote
import pl.treksoft.kvision.electron.fs.readFileSync
import pl.treksoft.kvision.electron.global
import pl.treksoft.kvision.html.button
import pl.treksoft.kvision.html.div
import pl.treksoft.kvision.i18n.DefaultI18nManager
import pl.treksoft.kvision.i18n.I18n
import pl.treksoft.kvision.i18n.I18n.tr
import pl.treksoft.kvision.panel.FlexAlignItems
import pl.treksoft.kvision.panel.hPanel
import pl.treksoft.kvision.panel.root
import pl.treksoft.kvision.panel.vPanel
import pl.treksoft.kvision.require
import pl.treksoft.kvision.startApplication
import pl.treksoft.kvision.utils.createInstance
import pl.treksoft.kvision.utils.px
import kotlin.browser.window

class ElectronApp : Application() {

    private var nativeWindow: BrowserWindow? = null

    override fun start() {

        I18n.manager =
            DefaultI18nManager(
                mapOf(
                    "en" to require("i18n/messages-en.json"),
                    "pl" to require("i18n/messages-pl.json")
                )
            )

        root("kvapp") {
            vPanel(alignItems = FlexAlignItems.CENTER, spacing = 30) {
                marginTop = 50.px
                fontSize = 30.px
                hPanel(spacing = 20) {
                    div(tr("Electron version"))
                    div(global.process.versions.electron)
                    button("open file") {
                        onClick {
                            val paths = remote.dialog.showOpenDialogSync(object : OpenDialogSyncOptions {}.apply {
                                title = "Open dialog title"
                            })
                            if (paths != undefined) {
                                val contents = readFileSync(paths[0], "utf8")
                                this.text = contents
                            }
                        }
                    }
                }
                hPanel(spacing = 20) {
                    div(tr("Chrome version"))
                    div(global.process.versions.chrome)
                }
            }
        }

        nativeWindow =
            remote.BrowserWindow.createInstance<BrowserWindow>(object : BrowserWindowConstructorOptions {}.apply {
                title = "Native window"
                width = 700
                height = 500
            })
        nativeWindow?.on("closed")
        { _ ->
            nativeWindow = null
        }
        window.onunload = {
            nativeWindow?.destroy()
            nativeWindow = null
            Unit
        }
    }

    override fun dispose(): Map<String, Any> {
        super.dispose()
        nativeWindow?.destroy()
        nativeWindow = null
        return mapOf()
    }
}

fun main() {
    startApplication(::ElectronApp)
}
```

**Comment by rjaros** [comment]
I'll also add `electron` NPM dependency to the KVision module. So it won't be necessary to include ` implementation(npm("electron"))` inside `build.gradle.kts`. It will also work with `umd` module kind again.

**Comment by rjaros** [comment]
I've forgot to add a reference to this issue in the commit message, so I just add a link here.
https://github.com/rjaros/kvision/commit/fc76ce268819451fb414abc0977ccfab5183b50a

**Comment by Lyeeedar** [comment]
Fantastic, thank you! Ill be looking forward to the next kvision release then. Happy for you to close this issue now :)

**Comment by rjaros** [comment]
If you (or anyone else!) would like to contribute, the generated code in the `pl.treksoft.kvision.electron` package would certainly need some cleanups (it gives me a lot of warnings), but I'm not able to get through all of it myself.

**Comment by rjaros** [comment]
Released with 3.11.2.

**Comment by rjaros** [comment]
As JetBrains added [experimental support for NodeJS API](https://github.com/Kotlin/kotlinx-nodejs) in Kotlin 1.4-rc, I plan to remove most of the NodeJS bindings I've created lately. They are generated from the same TypeScript declarations. Only the Electron bindings will be provided by KVision, and all NodeJS API will be available from JB bindings. Therefore some package changes will be required during migration.
